concretization: improve error messages for requirements

scheibelp opened the pull request:
This is intended to test out concretization error messages. Each test culminates in a concretization attempt that must fail (because the problem is over-constrained by design). 

Some of these tests add constraints to package definitions, and some add requirements to configuration. The error messages vary in quality but the tests are here to help experiment with tweaks to the `.lp` files to improve the messages. The tests do not include any enforcement of what appears in the error message: that has to be done manually, with something like `spack unit-test -k error_messages.py -s -v`.

Updates

* September 4 2025: all tests related to requirements are improved as of https://github.com/spack/spack/pull/45800/commits/ea7ab093839afc16ed1ce3ec3cdea8824bc82ee4
  * Theory: The concretizer has the freedom to choose a wrong version; the error weights for these type of messages was low, so it could short-circuit a cause chain by just assigning a wrong version
    * I've split the version error mechanism in two: arbitrary version selections that lead to version errors are penalized heavily. Version errors that result from other constraints are not.
  * A couple tests not related to requirements are generating bad error messages: `test_version_range_null`, `test_null_variant_for_requested_version`
* Jan. 22nd 2025: When this PR was first created, the error messages for requirements were particularly bad, but I improved them quite a bit by heavily encouraging requirement groups to apply at least one of their constraints, which appears to properly hook into the causal chaining logic in `error_messages.lp`.

TODOs

- [x] testing: the tests should include asserts rather than just printing the errors for manual inspection
  - (fixed as of September 8 in [0899dbd](https://github.com/spack/spack/pull/45800/commits/0899dbd8063448e17d06ebd92efbc1c63faa3fa6))

(old) ~All the tests here fail (on purpose)~

(old) ~All current tests added here do this by adding constraints to the package definition, but should probably also check mixing in config/command-line requirements.~

scheibelp committed:
simple example of a package conflict where the error message is useful

scheibelp committed:
naming convention for examples

scheibelp committed:
example with variants; this one also has a good error message

scheibelp committed:
add notes

scheibelp committed:
new test

scheibelp committed:
new example (w good error message)

scheibelp committed:
expanded last test case into something more interesting, with arguably less useful error message

scheibelp committed:
update doc

scheibelp committed:
more explanation

scheibelp committed:
style fix

scheibelp committed:
debugging/experimentation w/ error messages

scheibelp committed:
more notes

scheibelp committed:
Merge branch 'develop' into features/err-msg-tests

scheibelp committed:
test with requirement

scheibelp committed:
add explanatory comment

scheibelp committed:
example of terse error message

scheibelp committed:
comment

scheibelp committed:
demonstrate bad error message for require case whether it is encoded as config or in the package.py (what's strange is that sometimes, error messages from requirements are traceable)

scheibelp committed:
style fixes

scheibelp committed:
add explanatory comments

scheibelp committed:
minor experiment (revert)

scheibelp committed:
Merge branch 'develop' into features/err-msg-tests

scheibelp committed:
Merge branch 'develop' into features/err-msg-tests

scheibelp committed:
update license

scheibelp committed:
new import mechanism for packages

scheibelp committed:
style auto fix

scheibelp committed:
add comment on new behavior

scheibelp committed:
Merge branch 'develop' into features/err-msg-tests

scheibelp committed:
additional info

scheibelp committed:
rename to make it easier to run test by itself

scheibelp committed:
minor renames

scheibelp committed:
switch to problematic encoding for test

scheibelp committed:
rename

scheibelp committed:
show example of how minor tweak to requirements starts producing a bad error message

scheibelp committed:
intermediate work: show better handling of requirements

scheibelp committed:
remove unneeded

scheibelp committed:
update comment

scheibelp committed:
now that requirements force imposing one of their items, remove errors related to requirement group satisfaction

scheibelp committed:
swap to passing tests that print output (can add args to dump error message output for inspection if desired)

scheibelp committed:
auto style fix

scheibelp committed:
changes to error message logic affect customm error messages for requirements

scheibelp committed:
update test WRT new error message

scheibelp committed:
set upper bound for one_of

scheibelp commented:
I think I've figured out a way to improve error messages for failed concretizations related to requirements - that is on account of updates to `concretize.lp` (note: changes to `error_messages.lp` don't seem to have an effect either way - that was me trying to figure out the purpose of those statements). I've updated the PR description with that in mind. 

Because I wanted to exercise CI, I made it so all the tests added here pass, so if you want to actually see the improved error messages, you'd have to tell pytest to dump output for successful tests (also covered in the PR description) - ideally I can come up with some sensible way to enforce the quality of messages automatically that is not too fragile.

scheibelp committed:
merge from develop

scheibelp committed:
fix tests WRT new test module functionality (tmp_path, reintroduce create_test_repo)

scheibelp committed:
style fix

scheibelp committed:
fix a couple typos, and add a test related to external config

scheibelp committed:
style fix

scheibelp committed:
update comment based on new results after merging

scheibelp committed:
update comment to reflect recent results

scheibelp committed:
clarify comment

scheibelp committed:
add more comments based on experiments

scheibelp committed:
other style fix

scheibelp committed:
Merge branch 'develop' into features/err-msg-tests

scheibelp committed:
don't allow generating error messages for unsatisfied requirements (errors must be generated another way)

scheibelp committed:
in some cases the only error to report is a direct conflict between the user request and an explicitly encoded conflict: allow these to generate error messages, but heavily discourage them in favor of causal chaining

scheibelp committed:
add comments

scheibelp committed:
more comments

scheibelp committed:
error message in ...requirements_3 is now good, so rm old comment

scheibelp committed:
I have tracked down the difference between _2 and _3. I have no idea why this difference matters, but I at least have made it easy to demonstrate the difference

scheibelp committed:
style edit

scheibelp committed:
add note

scheibelp committed:
swap to bad msg example (so I can see if changing things improves it)

scheibelp committed:
more experimentation shows there seems to be a disconnect created for certain version specs

scheibelp committed:
notes on how to improve error message for _1 (it shouldnt be necessary to do that, but its worth knowing that it helps)

scheibelp committed:
Merge branch 'develop' into features/err-msg-tests

scheibelp committed:
improved message in existing test

scheibelp committed:
rm xfail (existing test for error messages now works)

scheibelp committed:
weighting against version errors improves error messages

scheibelp committed:
updated comments

scheibelp committed:
better result for existing version_error_messages test

scheibelp committed:
penalize version selection errors that arise only by choice rule; hack external error weights to show issue w/ test_errmsg_requirements_4

scheibelp committed:
allow choosing version from activated requirement (bypass penalty of arbitrary version selection)

scheibelp committed:
restore old weights

scheibelp committed:
augment requirement error message with activating condition (when there is one)

scheibelp committed:
can lower external error weight with 75730ecc46

scheibelp committed:
add some error-msg-checking logic to help determine whether messages are vague

scheibelp committed:
style fix

scheibelp committed:
add asymmetry to reduce redundancy in error

scheibelp committed:
improved comments and checking for test that is generating a bad error message

scheibelp committed:
satisfy flake

scheibelp commented on lib/spack/spack/solver/error_messages.lp:
```diff
@@ -41,22 +42,30 @@ condition_cause(Condition2, ID2, Condition1, ID1) :-
   condition_nodes(Trigger, TriggerNode, node(ID, Package)),
   trigger_node(Trigger, TriggerNode, node(ID2, Package2)),
-  attr(Name, node(ID, Package), A1, A2, A3),
+  %attr(Name, node(ID, Package), A1, A2, A3),
```
Note all the changes above this point in this file were me trying to figure out cause chaining and aren't essential to the PR (I was surprised I could effectively delete some of this and have the tests still pass).

becker33 replied:
I think that the `attr(` is redundant because we already have `condition_holds` and `impose_constraint`, so of course the attr must be true as well.

scheibelp committed:
merge from develop (minor semantic conflict, for now lowering weight of external error)

scheibelp committed:
I think generally speaking it makes sense to discourage picking no external at all: better to pick one and show the resulting conflicts from that

scheibelp committed:
merge from develop (minor sync issue)

becker33 commented on lib/spack/spack/solver/error_messages.lp:
```diff
@@ -106,7 +125,8 @@ error(0, "Cannot satisfy '{0}@{1}' and '{0}@{2}'", Package, Constraint1, Constra
      pkg_fact(Package, version_satisfies(Constraint1, Version)),
-     not pkg_fact(Package, version_satisfies(Constraint2, Version)).
+     not pkg_fact(Package, version_satisfies(Constraint2, Version)),
+     Cause1 < Cause2.
```
That's an important addition independent of the rest of the PR...

becker33 commented on lib/spack/spack/test/error_messages.py:
```diff
@@ -0,0 +1,554 @@
+
+
+@contextmanager
+def expect_failure_and_print(should_mention=None):
```
None of the working tests actually use the `should_mention` argument here -- should we be adding that to the tests to make them more specific? If not then we should probably remove it to keep this method simpler

scheibelp replied:
`test_null_variant_for_requested_version` uses this

EDIT: I think by "working" you mean "passing", `should_mention` is a reason the test fails, and it failing is useful to confirm that the error message is insufficient.

becker33 commented on lib/spack/spack/test/error_messages.py:
```diff
@@ -0,0 +1,554 @@
+    concretize_one("x4")
+
+    with expect_failure_and_print():
+        concretize_one("x1")
```
We should make these tests check more about the error message to make sure it stays good -- there's nothing here that actually tests for a good error message

scheibelp replied:
I oriented the checking toward making sure the error message is not bad, vs. checking that it is good.

e..g `test_diamond_with_pkg_conflict1` full error message is:

```
     1. Cannot satisfy 'x4@4.0' 4(4.1)
        required because x3 depends on x4@4.0 
     2. Cannot satisfy 'x4@4.0' 2(4.1)
     3. Cannot satisfy 'x4@4.1' and 'x4@4.0'
        required because x2 depends on x4@4.1 
        required because x3 depends on x4@4.0 
```

the methods `expect_failure_and_print`/`check_error` filter out all the generic error message content and make sure there is something left. Are you thinking we should in fact just compare against that string (maybe minus ordering)? I thought that might make the tests fragile.

scheibelp replied:
I could update that test to be like:

```
def test_diamond_with_pkg_conflict1(concretize_scope, test_repo):
    concretize_one("x2")
    concretize_one("x3")
    concretize_one("x4")

    important_points = [
        "x2 depends on x4@4.1",
        "x3 depends on x4@4.0",
    ]

    with expect_failure_and_print(should_mention=important_points):
        concretize_one("x1")
```

Is that the level of detail you are looking for in verification?

becker33 replied:
Yes, that's what I was envisioning.

scheibelp committed:
add defines to lp for body-referenced facts that are not always derived

scheibelp committed:
condense condition_cause rules and add explanations

also add positive assertions to error message test (make sure the
error message contains certain pieces of info)

scheibelp committed:
more positive checks in error messages

scheibelp committed:
auto style checks

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -291,14 +291,28 @@ pkg_fact(Package, version_declared(Version)) :- pkg_fact(Package, version_declar
+% version with a choice rule is penalized. This avoids the penalty
+% when the version comes from an activated requirement.
+% Oddly, does not seem to work if I relocate it to error_messages.lp
+{ attr("version", node(ID, Package), Version) } :-
+  requirement_group_member(ConditionID, Package, RequirementID),
+  activate_requirement(node(ID, Package), RequirementID),
+  pkg_fact(Package, condition_effect(ConditionID, EffectID)),
+  imposed_constraint(EffectID, "node_version_satisfies", Package, Version).
```
Some timing with requirements on would have been nice in this PR, so to know if there is any slow-down / speed-up to be expected.

scheibelp replied:
Apologies yes benchmarking would be useful: How do I do that? (i.e. how do I produce the before/after stats that you add to some PRs)?

This particular rule can be removed (although your comment likely does not apply to it specifically): I looked into it and this was originally added (in https://github.com/spack/spack/pull/45800/commits/75730ecc46d2de288522b2d944dc012fb0daf32a) for https://github.com/spack/spack/pull/45800/commits/344d77afdaa6429ba41fbcd086e714ccbe54b6a2, but an intervening merge removed that logic and it appears that as a result this doesn't actually improve error output for any of the tests (nor any that I can come up with)

alalazo replied:
> How do I do that? (i.e. how do I produce the before/after stats that you add to some PRs)?

We have a Spack extension here https://github.com/spack/spack-benchmark In the README you'll find the commands.
